src/details/ArborX_DetailsConcepts.hpp

```diff
@@ -60,6 +61,84 @@ struct has_centroid<
 {
 };
 
+template <typename Primitives, typename = void>
+struct has_access_traits : std::false_type
+{
+};
+
+template <typename Primitives>
+struct has_access_traits<
+    Primitives, std::void_t<typename Traits::Access<Primitives>::MemorySpace>>
+    : std::true_type
+{
+};
+
+template <typename Primitives, typename = void>
+struct has_memory_space : std::false_type
+{
+};
+
+template <typename Primitives>
+struct has_memory_space<
+    Primitives, std::enable_if_t<Kokkos::is_memory_space<
+                    typename Traits::Access<Primitives>::MemorySpace>::value>>
+    : std::true_type
+{
+};
+
+template <typename Primitives, typename = void>
+struct has_geometry_tag : std::false_type
+{
+};
+
+template <typename Primitives>
+struct has_geometry_tag<
+    Primitives,
+    std::enable_if_t<
+        std::is_same<typename Traits::Access<Primitives>::Tag,
+                     PointTag>::value ||
+        std::is_same<typename Traits::Access<Primitives>::Tag, BoxTag>::value>>
+    : std::true_type
+{
+};
+
+template <typename Primitives, typename = void>
+struct has_geometry_accessor : std::false_type
+{
+};
+
+template <typename Primitives>
+struct has_geometry_accessor<
+    Primitives,
+    std::enable_if_t<
+        (std::is_same<
+             std::decay_t<decltype(Traits::Access<Primitives>::get(
+                 std::declval<Primitives const &>(), std::declval<int>()))>,
+             Point>::value &&
+         std::is_same<typename Traits::Access<Primitives>::Tag,
+                      PointTag>::value) ||
+        (std::is_same<
+             std::decay_t<decltype(Traits::Access<Primitives>::get(
+                 std::declval<Primitives const &>(), std::declval<int>()))>,
+             Box>::value &&
+         std::is_same<typename Traits::Access<Primitives>::Tag,
+                      BoxTag>::value)>> : std::true_type
+{
+};
+
+template <typename Primitives, typename = void>
+struct has_size_accessor : std::false_type
+{
+};
+
+template <typename Primitives>
+struct has_size_accessor<
+    Primitives,
+    std::enable_if_t<std::is_integral<decltype(Traits::Access<Primitives>::size(
+        std::declval<Primitives const &>()))>::value>> : std::true_type
+{
+};
+
```

Comment:
In principle these could be templated on the access traits directly.  I don't have a strong opinion on this.  It might be subject to change when we introduce access traits for the predicates.

Reply:
If we only use it on `Traits::Access` I would certainly prefer that.

Reply:
You are commenting on an old revision.